zookeeper-api/src/main/java/org/apache/helix/zookeeper/util/HttpRoutingDataReader.java

```diff
@@ -0,0 +1,134 @@
+package org.apache.helix.zookeeper.util;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+import com.jcabi.aspects.RetryOnFailure;
+import org.apache.helix.msdcommon.constant.MetadataStoreRoutingConstants;
+import org.apache.helix.msdcommon.datamodel.MetadataStoreRoutingData;
+import org.apache.helix.msdcommon.datamodel.TrieRoutingData;
+import org.apache.helix.msdcommon.exception.InvalidRoutingDataException;
+import org.apache.http.HttpEntity;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.util.EntityUtils;
+import org.codehaus.jackson.map.ObjectMapper;
+
+
+public class HttpRoutingDataReader {
+  private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+  private static volatile Map<String, List<String>> _rawRoutingData;
+  private static volatile MetadataStoreRoutingData _metadataStoreRoutingData;
+
+  /**
+   * This class is a Singleton.
+   */
+  private HttpRoutingDataReader() {
+  }
+
+  /**
+   * Fetches routing data from the data source via HTTP.
+   * @return a mapping from "metadata store realm addresses" to lists of
+   * "metadata store sharding keys", where the sharding keys in a value list all route to
+   * the realm address in the key disallows a meaningful mapping to be returned
+   */
+  @RetryOnFailure(attempts = 3, delay = 1, unit = TimeUnit.SECONDS)
```

Comment:
What is the benefit of introducing new lib: jcabi. I think HttpClients' already provides retry mechanism: `HttpRequestInterceptor` or `setRetryHandler(new DefaultHttpRequestRetryHandler())`

Reply:
The benefit of using jcabi is that it allows us to retry in a more succinct way. Defining a retry handler every time would incur repeated code.

I want to use https://http.jcabi.com/ (Fluent HttpClient) over Apache's at some point once we refactor all HttpClients out into a new module because it provides useful annotations for caching and logging, but this is more like an ad-hoc retry here. Let me remove this library import. I'll just use the retry handler. 

Thanks for the suggestion!